keps/sig-cli/0031-kustomize-integration.md

```diff
@@ -0,0 +1,230 @@
+
+### Goals
+
+Integrate kustomize with kubectl so that kubectl can recognize kustomization directories and expand resources from kustomization.yaml before running kubectl subcommands. This integration should be transparent. It doesn't change kubectl UX. This integration should also be backward compatible. For non kustomization directories, kubectl behaves the same as current. The integration shouldn't have any impact on those parts.
+
+
+### Non-Goals
+- provide an editing functionality of kustomization.yaml from kubectl
+- further integration with other kubectl flags
+
+## Kustomize Introduction
+
+Kustomize has following subcommands:
+- build
+- edit, edit also has subcommands
+   - Set 
+       - imagetag
+       - namespace
+       - nameprefix
+   - Add
+       - base
+       - resource
+       - patch
+       - label
+       - annotation
+       - configmap
+- config
+- version
+- help
+
+`edit` and `build` are most commonly used subcommands. 
+
+`edit` is to modify the fields in `kustomization.yaml`. A `kustomization.yaml` includes configurations that are consumed by Kustomize. Here is an example of `kustomization.yaml` file.
+
+`build` is to perform a set of pre-processing transformations on the resources inside one kustomization. Those transformations include:
+- Get objects from the base
+- Apply patches
+- Add name prefix to all resources
+- Add common label and annotation to all resources
+- Replace imageTag is specified
+- Update objects’ names where they are referenced
+- Resolve variables and substitute them
+
+```
+namePrefix: alices-
+
+commonAnnotations:
+ oncallPager: 800-555-1212
+
+configMapGenerator:
+- name: myJavaServerEnvVars
+ literals:   
+ - JAVA_HOME=/opt/java/jdk
+ - JAVA_TOOL_OPTIONS=-agentlib:hprof
+
+secretGenerator:
+- name: app-sec
+ commands:
+   username: "echo admin"
+   password: "echo secret"
+```
+The build output of this sample kustomizaiton.yaml file is
+```
+apiVersion: v1
+data:
+ JAVA_HOME: /opt/java/jdk
+ JAVA_TOOL_OPTIONS: -agentlib:hprof
+kind: ConfigMap
+metadata:
+ annotations:
+   oncallPager: 800-555-1212
+ name: alices-myJavaServerEnvVars-7bc9c27cmf
+---
+apiVersion: v1
+data:
+ password: c2VjcmV0Cg==
+ username: YWRtaW4K
+kind: Secret
+metadata:
+ annotations:
+   oncallPager: 800-555-1212
+ name: alices-app-sec-c7c5tbh526
+type: Opaque
+```
+
+## Proposal
+
+### UX
+
+When apply, get or delete is run on a directory, check if it contains a kustomization.yaml file. If there is, apply, get or delete the output of kustomize build. Kubectl behaves the same as current for directories without kustomization.yaml.
```

Comment:
What about for other commands that take directories? They should be processed by kustomize or fail.